src/main/java/com/faforever/api/map/MapService.java

```diff
@@ -0,0 +1,131 @@
+    try (ZipInputStream zipInputStream = new ZipInputStream(new BufferedInputStream(Files.newInputStream(tmpFile)))) {
+      Unzipper unzipper = Unzipper.from(zipInputStream);
+      unzipper.to(baseDir);
+      unzipper.unzip();
+    }
+
+    // read from Lua File
+    Path scenarioLuaPath = noCatch(() -> list(baseDir))
+        .filter(myFile -> myFile.toString().endsWith("_scenario.lua"))
+        .findFirst()
+        .orElseThrow(() -> new IllegalStateException("Map folder does not contain a *_scenario.lua: " + baseDir.toAbsolutePath()));
+
+    LuaValue luaRoot = noCatch(() -> loadFile(scenarioLuaPath), IllegalStateException.class);
+    LuaValue scenarioInfo = luaRoot.get("ScenarioInfo");
+
+
+    Optional<Map> mapEntity = mapRepository.findOneByDisplayName(scenarioInfo.get("name").toString());
+    if (mapEntity.isPresent() && mapEntity.get().getAuthor().getId() != author.getId()) {
+      throw new ValidationException("Your are not the author of the map");
+    }
+    if (mapEntity.isPresent() && mapEntity.get().getVersions().stream()
+        .anyMatch(mapVersion -> mapVersion.getVersion() == scenarioInfo.get("map_version").toint())) {
+      throw new ValidationException("Map Version already exists");
+    }
+
+    Map map = mapEntity.isPresent() ? mapEntity.get() : new Map();
+    map.setDisplayName(scenarioInfo.get("name").toString());
+    map.setMapType(scenarioInfo.get("type").tojstring());
+    map.setBattleType(scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("name").tojstring());
+    map.setAuthor(author);
+
+    // try to save entity to db to trigger validation
+    LuaValue size = scenarioInfo.get("size");
+    MapVersion version = new MapVersion();
+    version.setFilename(mapFilename);
+    version.setDescription(scenarioInfo.get("description").tojstring().replaceAll("<LOC .*?>", ""));
+    version.setWidth((int) (size.get(1).toint() / MAP_SIZE_FACTOR));
+    version.setHeight((int) (size.get(2).toint() / MAP_SIZE_FACTOR));
+    version.setHidden(false);
+    version.setRanked(false);
+    version.setMaxPlayers(scenarioInfo.get("Configurations").get("standard").get("teams").get(1).get("armies").length());
+    version.setVersion(scenarioInfo.get("map_version").toint());
+
+    map.getVersions().add(version);
+    version.setMap(map);
+    mapRepository.save(map);
+
+    // generate preview
+    String previewFilename = FilenameUtils.removeExtension(mapFilename) + ".png";
+    generateImage(Paths.get(
+        fafApiProperties.getMap().getMapPreviewPathSmall(), previewFilename),
+        baseDir,
+        fafApiProperties.getMap().getPreviewSizeSmall());
+
+    generateImage(Paths.get(
+        fafApiProperties.getMap().getMapPreviewPathLarge(), previewFilename),
+        baseDir,
+        fafApiProperties.getMap().getPreviewSizeLarge());
+
+
+    // move to final path
+    Files.createDirectories(finalPath.getParent());
+    FileCopyUtils.copy(mapData, finalPath.toFile());
```

Comment:
Why do we write this again? Also, in the existing API code we replace some file contents and re-package the map file, which I'm missing here (but I assume you're PR is WIP, in which case you should add FIXME's however, so everyone knows in case you get hit by a bus ;))